File: widgets/symphony_data_map/src/widget.ts
// For licensing see accompanying LICENSE file.
// Copyright (C) 2023 Apple Inc. All Rights Reserved.

import { MODULE_NAME, MODULE_VERSION } from './version';
import {
  DOMWidgetModel,
  DOMWidgetView,
  ISerializers,
} from '@jupyter-widgets/base';

import { serializers, JupyterWidget } from '@apple/symphony-lib';

import SymphonyDataMap from './SymphonyDataMap.svelte';

export class SymphonyDataMapModel extends DOMWidgetModel {
  defaults() {
    return {
      ...super.defaults(),
      _model_name: SymphonyDataMapModel.model_name,
      _model_module: SymphonyDataMapModel.model_module,
      _model_module_version: SymphonyDataMapModel.model_module_version,
      _view_name: SymphonyDataMapModel.view_name,
      _view_module: SymphonyDataMapModel.view_module,
      _view_module_version: SymphonyDataMapModel.view_module_version,
    };
  }
  static serializers: ISerializers = serializers;

  static model_name = 'SymphonyDataMapModel';
  static model_module = MODULE_NAME;
  static model_module_version = MODULE_VERSION;
  static view_name = 'SymphonyDataMapView'; // Set to null if no view
  static view_module = MODULE_NAME; // Set to null if no view
  static view_module_version = MODULE_VERSION;
}

export class SymphonyDataMapView extends DOMWidgetView {
  render() {
    new JupyterWidget({
      target: this.el,
      props: {
        model: this.model,
        widget: SymphonyDataMap,
      },
    });
  }
}

File: widgets/symphony_data_map/symphony_data_map/widget.py
# For licensing see accompanying LICENSE file.
# Copyright (C) 2023 Apple Inc. All Rights Reserved.

from pathlib import Path

import symphony_ui
from traitlets import Unicode

from ._frontend import module_name, module_version
from ._default_codes import codes


class SymphonyDataMap(symphony_ui.SymphonyWidget):
    _model_name = Unicode(
        'SymphonyDataMapModel').tag(sync=True)
    _view_name = Unicode(
        'SymphonyDataMapView').tag(sync=True)
    _model_module = Unicode(module_name).tag(sync=True)
    _model_module_version = Unicode(module_version).tag(sync=True)
    _view_module = Unicode(module_name).tag(sync=True)
    _view_module_version = Unicode(module_version).tag(sync=True)
    name = 'SymphonyDataMap'
    description = "A Symphony component for visualizing data on a map"

    def __init__(self,
                 width: str = 'XXL',
                 height: str = 'M',
                 page: str = 'Data Map',
                 projection: str = 'mercator',
                 id_map: dict = codes,
                 feature: str = 'countries',
                 id_column: str = 'country',
                 map_url: str = 'https://raw.githubusercontent.com/vega/datalib/master/test/data/world-110m.json',
                 ** kwargs
                 ):
        """A Symphony component for visualizing data on a map

        Parameters
        ----------
        width : str, optional
            By default "XXL".
        height : str, optional
            By default "M".
        page : str, optional
            Which page of the dashboard to show the widget on, by default "SymphonyDataMap"

        Returns
        -------
        SymphonyDataMap
        """
        super().__init__(**kwargs)
        self.width: str = width
        self.height: str = height
        self.page: str = page
        self.spec = symphony_ui.DataMapSpec(
            width=width,
            height=height,
            page=page,
            name=self.name,
            description=self.description,
            projection=projection,
            id_map=id_map,
            feature=feature,
            id_column=id_column,
            map_url=map_url
        )

    def js_path(self):
        app_files_path = Path(
            (Path(__file__).parent), 'standalone', 'widgets')
        js_path = Path(app_files_path, 'SymphonyDataMap.js')
        map_path = Path(
            app_files_path, 'SymphonyDataMap.js.map')
        return js_path, map_path
